metadb/server.go

```diff
@@ -0,0 +1,337 @@
+			return errors.Errorf("store log not applied within timeout: %s", err.Error())
+		}
+	} else {
+		log.WithContext(ctx).Info("not waiting for logs to be applied")
+	}
+
+	return nil
+}
+
+// open the auth file, and returns a credential store instance
+func (s *Service) credentialStore() (*auth.CredentialsStore, error) {
+	if s.config.AuthFile == "" {
+		return nil, nil
+	}
+
+	file, err := os.Open(s.config.AuthFile)
+	if err != nil {
+		return nil, errors.Errorf("open authentication file: %v", err)
+	}
+
+	store := auth.NewCredentialsStore()
+	if store.Load(file); err != nil {
+		return nil, errors.Errorf("store load: %v", err)
+	}
+	return store, nil
+}
+
+// start the http server
+func (s *Service) startHTTPServer(dbStore *store.Store, cs *cluster.Service) error {
+	// load the credential store
+	cred, err := s.credentialStore()
+	if err != nil {
+		return errors.Errorf("load credentail store: %v", err)
+	}
+
+	var server *httpd.Service
+	// create http server and load authentication information if required
+	if cred != nil {
+		server = httpd.New(s.config.HTTPAddress, dbStore, cs, cred)
+	} else {
+		server = httpd.New(s.config.HTTPAddress, dbStore, cs, nil)
+	}
+	server.CertFile = s.config.X509Cert
+	server.KeyFile = s.config.X509Key
+	server.TLS1011 = s.config.TLS1011
+
+	// start the http server
+	return server.Start()
+}
+
+// start a mux for rqlite node
+func (s *Service) startNodeMux(ctx context.Context, ln net.Listener) (*tcp.Mux, error) {
+	var adv net.Addr
+	var err error
+	if s.config.RaftAdvertiseAddress != "" {
+		adv, err = net.ResolveTCPAddr("tcp", s.config.RaftAdvertiseAddress)
+		if err != nil {
+			return nil, errors.Errorf("resolve advertise address %s: %v", s.config.RaftAdvertiseAddress, err)
+		}
+	}
+
+	var mux *tcp.Mux
+	if s.config.NodeEncrypt {
+		log.WithContext(ctx).Infof("enabling node-to-node encryption with cert: %s, key: %s", s.config.NodeX509Cert, s.config.NodeX509Key)
+		mux, err = tcp.NewTLSMux(ln, adv, s.config.NodeX509Cert, s.config.NodeX509Key, s.config.NodeX509CACert)
+	} else {
+		mux, err = tcp.NewMux(ln, adv)
+	}
+	if err != nil {
+		return nil, errors.Errorf("create node-to-node mux: %s", err.Error())
+	}
+	mux.InsecureSkipVerify = s.config.NodeNoVerify
+
+	go mux.Serve()
+
+	return mux, nil
+}
+
+// start the cluster server
+func (s *Service) startClusterService(tn cluster.Transport) (*cluster.Service, error) {
+	c := cluster.New(tn)
+
+	apiAddr := s.config.HTTPAddress
+	if s.config.HTTPAdvertiseAddress != "" {
+		apiAddr = s.config.HTTPAdvertiseAddress
+	}
+	c.SetAPIAddr(apiAddr)
+	// conditions met for a HTTPS API
+	c.EnableHTTPS(s.config.X509Cert != "" && s.config.X509Key != "")
+
+	// open the cluster service
+	if err := c.Open(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// start the rqlite server, and try to join rqlite cluster if the join addresses is not empty
+func (s *Service) startServer(ctx context.Context) error {
```

Comment:
too long function, better to split it by meaning

Reply:
done